### PR TITLE
[stable/mongodb] - Refactor MongoDB chart to support replica set configurations

### PR DESCRIPTION
**Pull request by tompizmor**
**What this PR does / why we need it**:
This PR adds support to the stable/mongodb chart to be deployed as a replica set configuring primary, secondary and arbiter nodes.

Other changes:
- When deploying in replica-set mode, it uses stateful-sets with headless-svc and volumeClaimTemplates for persistence.
- When using replica-set mode, pod-disruption-budgets are configurable too 
- It allows to overwrite the default config file with a config map
- replica set parameters are configurable through values.yaml
- Two `values.yaml` availables, `values.yaml` and `values-production.yaml`

**Special notes for your reviewer**:
Please **do not merge** until I say the contrary as I need to make publish the new version of the docker image.

### CODE REVIEW

File: stable/mongodb/values.yaml [hunk 1]
```diff
@@ -1,15 +1,20 @@
-## Bitnami MongoDB image version
-## ref: https://hub.docker.com/r/bitnami/mongodb/tags/
-##
 image:
+  ## Bitnami MongoDB registry
+  ##
   registry: docker.io
-  repository: bitnami/mongodb
```
Comment (juan131) [inline stable/mongodb/values.yaml]: Why did you change this parameter name? By the way, it's not aligned with the `README.md`

File: stable/mongodb/values.yaml [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
+  ##
+  name: bitnami/mongodb
+  ## Bitnami MongoDB image tag
+  ## ref: https://hub.docker.com/r/bitnami/mongodb/tags/
+  ##
   tag: 3.6.4
+
   ## Specify a imagePullPolicy
-  ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
+  ## 'Always' if imageTag is 'latest', else set to 'IfNotPresent'
```
Comment (juan131) [inline stable/mongodb/values.yaml]: `imageTag` -> `image.tag`

File: stable/mongodb/values.yaml [hunk 3]
```diff
@@ -42,15 +48,67 @@ usePassword: true
 ##  - "--wiredTigerCacheSizeGB=2"
 mongodbExtraFlags: []
 
+## Kubernetes service type
 service:
-  ## Kubernetes service type
```
Comment (juan131) [inline stable/mongodb/values.yaml:46]: Why did you move this line?

File: stable/mongodb/values-production.yaml [hunk 4]
```diff
@@ -0,0 +1,184 @@
+image:
+  ## Bitnami MongoDB registry
+  ##
+  registry: docker.io
+  ## Bitnami MongoDB image name
+  ##
+  name: bitnami/mongodb
```
Comment (juan131) [inline stable/mongodb/values-production.yaml]: Why did you change this parameter name? By the way, it's not aligned with the README.md

File: stable/mongodb/values-production.yaml [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  ## Bitnami MongoDB image tag
+  ## ref: https://hub.docker.com/r/bitnami/mongodb/tags/
+  ##
+  tag: 3.6.4
+
+  ## Specify a imagePullPolicy
+  ## 'Always' if imageTag is 'latest', else set to 'IfNotPresent'
```
Comment (juan131) [inline stable/mongodb/values-production.yaml]: `imageTag` -> `image.tag`

File: stable/mongodb/values-production.yaml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+## MongoDB additional command line flags
+##
+## Can be used to specify command line flags, for example:
+##
+## mongodbExtraFlags:
+##  - "--wiredTigerCacheSizeGB=2"
+mongodbExtraFlags: []
+
+## Kubernetes service type
```
Comment (juan131) [inline stable/mongodb/values-production.yaml:50]: Why did you move this line?

File: stable/mongodb/templates/deployment-standalone.yaml [hunk 7]
```diff
@@ -26,9 +27,11 @@ spec:
       {{- end }}
       containers:
       - name: {{ template "mongodb.fullname" . }}
-        image: {{ template "mongodb.image" . }}
+        image: "{{ .Values.image.registry }}/{{ .Values.image.name }}:{{ .Values.image.tag }}"
```
Comment (juan131) [inline stable/mongodb/templates/deployment-standalone.yaml]: Define a function on *_helpers.tpl* for this and use it.

File: stable/mongodb/templates/deployment-standalone.yaml [hunk 8]
```diff
@@ -26,9 +27,11 @@ spec:
...
         imagePullPolicy: {{ .Values.image.pullPolicy | quote }}
         env:
+        - name: NAMI_LOG_LEVEL
```
Comment (juan131) [inline stable/mongodb/templates/deployment-standalone.yaml]: Please remove this

File: stable/mongodb/templates/headless-svc-rs.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+kind: Service
+metadata:
+  name: {{ template "mongodb.fullname" . }}-headless
+  labels:
+    app: {{ template "mongodb.name" . }}
+    chart: {{ template "mongodb.chart" . }}
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  type: ClusterIP
```
Comment (juan131) [inline stable/mongodb/templates/headless-svc-rs.yaml:12]: Should this be configurable?

**Review by juan131 — CHANGES_REQUESTED** [review]

File: stable/mongodb/templates/headless-svc-rs.yaml (hunk 9 above)
Comment (tompizmor) [reply stable/mongodb/templates/headless-svc-rs.yaml:12]: nop, this one should be ClusterIp and ClusterIp: Nono

File: stable/mongodb/templates/deployment-standalone.yaml (hunk 8 above)
Comment (tompizmor) [reply stable/mongodb/templates/deployment-standalone.yaml]: ok

File: stable/mongodb/values.yaml (hunk 1 above)
Comment (tompizmor) [reply stable/mongodb/values.yaml]: human mistake!

File: stable/mongodb/templates/deployment-standalone.yaml (hunk 7 above)
Comment (tompizmor) [reply stable/mongodb/templates/deployment-standalone.yaml]: ok

File: stable/mongodb/values-production.yaml (hunk 6 above)
Comment (tompizmor) [reply stable/mongodb/values-production.yaml:50]: I see it fine

**Comment by unguiculus** [comment]
Why do we need this? There's already https://github.com/kubernetes/charts/tree/master/stable/mongodb-replicaset.

cc @prydonius

**Comment by tompizmor** [comment]
@unguiculus We have done similar PRs for Rabbitmq (https://github.com/kubernetes/charts/pull/4591) and Redis (https://github.com/kubernetes/charts/pull/4662) too.

The configuration for mongodb and mongodb-replicaset are a bit different. 

* mongodb-replicaset deploys X mongodb  replicas and one of the is the master (you need to find which one). The mongodb chart has a fixed stateful set for the master. 
* mongodb-replicaset chart does not use arbiter nodes while mongodb chart does
* The replication is disabled by default in the mongodb chart, but it can be easily enabled, so it is easy to move applications from a single mongodb node to a replicaset scenario.

cc/ @prydonius @sameersbn what do you think?

File: stable/mongodb/README.md [hunk 10]
```diff
@@ -43,36 +43,52 @@ The command removes all the Kubernetes components associated with the chart and
...
+|         Parameter                       |             Description                                                                      |                         Default                          |
+|-----------------------------------------|----------------------------------------------------------------------------------------------|----------------------------------------------------------|
+| `image.registry`                        | MongoDB image registry                                                                       | `docker.io`                                              |
+| `image.repository`                      | MongoDB Image name                                                                           | `bitnami/mongodb`                                        |
+| `image.tag`                             | MongoDB Image tag                                                                            | `{VERSION}`                                              |
+| `image.pullPolicy`                      | Image pull policy                                                                            | `Always` if `imageTag` is `latest`, else `IfNotPresent`  |
```
Comment (sameersbn) [inline stable/mongodb/README.md]: The default value column should just say `Always`

File: stable/mongodb/templates/statefulset-arbiter-rs.yaml [hunk 11]
```diff
@@ -0,0 +1,86 @@
+{{- if .Values.replicaSet.enabled }}
+apiVersion: apps/v1
+kind: StatefulSet
```
Comment (sameersbn) [inline stable/mongodb/templates/statefulset-arbiter-rs.yaml:3]: the arbiter nodes do not save any data, do we really need to use a statefulset here?

File: stable/mongodb/values-production.yaml [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+service:
+  type: ClusterIP
+
+  ## Specify the nodePort value for the LoadBalancer and NodePort service types.
+  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
+  ##
+  # nodePort:
+
+port: 27017
```
Comment (sameersbn) [inline stable/mongodb/values-production.yaml]: should we move `port` to `service.port`?

File: stable/mongodb/values-production.yaml [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+  ## Bitnami MongoDB image name
+  ##
+  repository: bitnami/mongodb
+  ## Bitnami MongoDB image tag
+  ## ref: https://hub.docker.com/r/bitnami/mongodb/tags/
+  ##
+  tag: 3.6.4
+
+  ## Specify a imagePullPolicy
+  ## 'Always' if image.tag is 'latest', else set to 'IfNotPresent'
```
Comment (sameersbn) [inline stable/mongodb/values-production.yaml]: remove this comment

File: stable/mongodb/values-production.yaml [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  ##
+  # storageClass: "-"
+  accessModes:
+    - ReadWriteOnce
+  size: 8Gi
+  annotations: {}
+
+## Configure extra options for liveness and readiness probes
+## ref: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes)
+livenessProbe:
```
Comment (sameersbn) [inline stable/mongodb/values-production.yaml:136]: we should be able to completely disable the probes

File: stable/mongodb/README.md [hunk 15]
```diff
@@ -43,36 +43,52 @@ The command removes all the Kubernetes components associated with the chart and
...
+| `readinessProbe.timeoutSeconds`         | When the probe times out                                                                     | `5`                                                      |
+| `readinessProbe.failureThreshold`       | Minimum consecutive failures for the probe to be considered failed after having succeeded.   | `6`                                                      |
+| `readinessProbe.successThreshold`       | Minimum consecutive successes for the probe to be considered successful after having failed. | `1`                                                      |
+| `configmap`                             | MongoDB configuration file to be used                                                        | `nil`                                                    |
 
 The above parameters map to the env variables defined in [bitnami/mongodb](http://github.com/bitnami/bitnami-docker-mongodb). For more information please refer to the [bitnami/mongodb](http://github.com/bitnami/bitnami-docker-mongodb) image documentation.
```
Comment (sameersbn) [inline stable/mongodb/README.md]: we should remove this line. It does not make sense anymore.
We need to add a section on deploying the chart with values-production.yaml configuration (see https://github.com/kubernetes/charts/tree/master/stable/wordpress#production-and-horizontal-scaling)

**Review by sameersbn — CHANGES_REQUESTED** [review]

File: stable/mongodb/README.md (hunk 10 above)
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/templates/statefulset-arbiter-rs.yaml (hunk 11 above)
Comment (tompizmor) [reply stable/mongodb/templates/statefulset-arbiter-rs.yaml:3]: we needed so the network identifier is stable.

File: stable/mongodb/values-production.yaml (hunk 12 above)
Comment (tompizmor) [reply stable/mongodb/values-production.yaml]: done

File: stable/mongodb/values-production.yaml (hunk 13 above)
Comment (tompizmor) [reply stable/mongodb/values-production.yaml]: done

File: stable/mongodb/values-production.yaml (hunk 14 above)
Comment (tompizmor) [reply stable/mongodb/values-production.yaml:136]: done

File: stable/mongodb/README.md (hunk 15 above)
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md [hunk 16]
```diff
@@ -112,10 +110,20 @@ $ helm install --name my-release -f values.yaml stable/mongodb
...
-In order to set up a MongoDB replica set with primary, secondary and arbiter nodes you can use the values-production.yaml file.
+You can start the MongoDB chart in replica set mode with the following command:
 
 ```bash
-$ helm install --name my-release -f values-production.yaml stable/mongodb
+$ helm install --name my-release stable/mongodb --set replication.enabled=true
+```
+## Production settings and horizontal scaling
+
+The following repo contains the recommended production settings for MongoDB server in an alternative [values file](values-production.yaml). Please read carefully the comments in the values-production.yaml file to set up your environment appropriately.
```
Comment (sameersbn) [inline stable/mongodb/README.md]: I would change this to,

The [values-production.yaml](values-production.yaml) file consists a configuration to deploy a scalable MongoDB deployment for production environments. We recommend that you base your production configuration on this template and adjust the parameters appropriately.

```bash
$ curl -O https://raw.githubusercontent.com/kubernetes/charts/master/stable/mongodb/production-values.yaml
$ helm install --name my-release -f values-production.yaml stable/mongodb
```

File: stable/mongodb/README.md [hunk 17]
```diff
@@ -112,10 +110,20 @@ $ helm install --name my-release -f values.yaml stable/mongodb
...
+
+To horizontally scale this chart, first download the [values-production.yaml](values-production.yaml) file to your local folder, then scale the secondary nodes stateful set:
```
Comment (sameersbn) [inline stable/mongodb/README.md]: `stateful set` -> `statefulset`

File: stable/mongodb/README.md [hunk 18]
```diff
@@ -112,10 +110,20 @@ $ helm install --name my-release -f values.yaml stable/mongodb
...
+
+```console 
+$ helm install --name my-release -f ./values-production.yaml stable/mongodb
+$ kubectl scale statefulset my-mongodb-secondaries-statefulset --replicas=3
```
Comment (sameersbn) [inline stable/mongodb/README.md]: I think name of the statefulset in the command is incorrect. Should probably be 

```bash
$ kubectl scale statefulset my-release-mongodb-secondary --replicas=3 
```

Ideally we should be documenting the `helm upgrade` way but I understand that the use of random passwords is the issue here

**Review by sameersbn — CHANGES_REQUESTED** [review]

File: stable/mongodb/README.md (hunk 16 above)
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md (hunk 17 above)
Comment (tompizmor) [reply stable/mongodb/README.md]: done

File: stable/mongodb/README.md (hunk 18 above)
Comment (tompizmor) [reply stable/mongodb/README.md]: fixed

**Review by sameersbn — APPROVED** [review]

**Comment by sameersbn** [comment]
/approve /ok-to-test

Please wait for approval by @unguiculus before the merge.

**Comment by tompizmor** [comment]
ping @unguiculus

**Comment by unguiculus** [comment]
I'm not going to block this but was hoping @prydonius would jump in. We had a similar discussion on another chart. I think it was Rabbitmq.

**Comment by prydonius** [comment]
I think this is okay, but we should outline the differences you mentioned in https://github.com/kubernetes/charts/pull/5578#issuecomment-392482527 in the Chart's README in a section about enabling high-availability.

**Comment by tompizmor** [comment]
@prydonius, updated!

File: stable/mongodb/README.md [hunk 19]
```diff
@@ -106,18 +104,37 @@ Alternatively, a YAML file that specifies the values for the parameters can be p
 
 ```bash
 $ helm install --name my-release -f values.yaml stable/mongodb
-```
```
Comment (juan131) [inline stable/mongodb/README.md]: Why has this being removed?

Comment (tompizmor) [reply stable/mongodb/README.md]: mistake, fixed

**Review by juan131 — APPROVED** [review]
Thanks!

File: stable/mongodb/README.md [hunk 20]
```diff
@@ -92,9 +105,36 @@ Alternatively, a YAML file that specifies the values for the parameters can be p
...
 ```
-
 > **Tip**: You can use the default [values.yaml](values.yaml)
 
+## Replication
+
+You can start the MongoDB chart in replica set mode with the following command:
+
+```bash
+$ helm install --name my-release stable/mongodb --set replication.enabled=true
```
Comment (sabbour) [inline stable/mongodb/README.md:115]: @tompizmor in the README, the command to install the Helm chart with replication is incorrect. `replication.enabled=true` should be `replicaSet.enabled=true`.

**Pull request merged**